### PR TITLE
[css-grid] Resolve and set the row start, non-auto margins for subgrids with baseline aligned items.

### PR DESCRIPTION
**Pull request by sammygill**
#### d28a4da714e84e52e8c59db1b828d7f601c34036
<pre>
[css-grid] Resolve and set the row start, non-auto margins for subgrids with baseline aligned items.
<a href="https://bugs.webkit.org/show_bug.cgi?id=262508">https://bugs.webkit.org/show_bug.cgi?id=262508</a>
rdar://problem/116369419

Reviewed by Matt Woodrow.

Note: This patch focuses on when we are performing baseline alignment
in the block direction of a grid (i.e. align-self/align-items).

When subgrids baseline align their items as part of alignment in the
outer grid, the sum of the subgrid&apos;s margin, border, and padding are
treated as an extra layer of margin for the grid item. Since an item&apos;s
margin is part of what determines the ascent value for the purposes of
baseline alignment, we must have that information when it gets added
into a baseline alignment context.

This process occurs at the end of the GridTrackSizingAlgorithm&apos;s setup
phase by the call to computeBaselineAlignmentContext, but the issue is
that by the time we call computeBaselineAlignmentContext none of the
subgrids have their margins resolved as part of their m_marginBox. This
information gets set during the actual grid track sizing algorithm when
the subgrid goes through layout as we attempt to compute the min content
size of the subgridded item.

When we call cacheBaselineAlignedChildren on the &quot;root&quot; grid (the grid
that subgrids will share tracks with), we can keep track of a bool that
is initially set to true to represent this state in which we want to
cache row subgrids. After we cache a baseline aligned item in
cacheBaselineAlignedItem we can then check to see if this flag is true
and also cache the item&apos;s grid if it is a row subgrid of the root grid.

Later on we will go through these subgrids and resolve their
margins just before we start creating the baseline alignment context
for the GridColumnAxis. We will only do this as part of the first/second
iteration of the row track sizing since we will just have determined the
size of the columns in the previous step. The sizes of the columns are
needed so that we can get the available width from the grid area the
subgrid is contained in, which may be needed to resolve its margins in
certain cases.

The affected tests below are brought more in like with their non-iOS
counterparts which share the same expected results. The discrepancy
between the two platforms is because on macOS, for example, an
additional layout occurs via a call to LocalFrameView::adjustViewSize
that does not occur when running the tests on iOS. The second layout
then uses the margin box information for the subgrids that was set
previously.

* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-grid/subgrid/align-self-baseline-with-subgrid-mbp-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-005-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-006-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-007-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-008-expected.txt:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::computeGridSpanSize):
(WebCore::GridTrackSizingAlgorithm::gridAreaBreadthForChild const):
(WebCore::GridTrackSizingAlgorithm::cacheBaselineAlignedItem):
(WebCore::GridTrackSizingAlgorithm::setup):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::cacheBaselineAlignedChildren):
(WebCore::RenderGrid::computeIntrinsicLogicalWidths const):
(WebCore::RenderGrid::isSubgridOf const):
(WebCore::RenderGrid::isSubgridOf): Deleted.
* Source/WebCore/rendering/RenderGrid.h:

Canonical link: <a href="https://commits.webkit.org/269424@main">https://commits.webkit.org/269424@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/0c6bf8c0684e1732fe81acb3f0d233c84cbe1233

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21956 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22173 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23023 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23838 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20328 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22201 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26421 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22485 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21408 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22186 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21823 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19042 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24690 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18948 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19896 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26161 "Passed tests") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19974 "Found 1 new API test failure: TestWebKitAPI.WebArchive.SaveResourcesBlobURL (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20118 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24026 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20552 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17606 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19905 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5366 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24110 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20503 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp [hunk 1]
```diff
@@ -637,6 +637,14 @@ std::optional<LayoutUnit> GridTrackSizingAlgorithm::estimatedGridAreaBreadthForC
     return gridAreaSize;
 }
 
+static LayoutUnit computeGridAreaSize(const Vector<GridTrack>& tracks, const GridSpan& gridArea, const std::optional<LayoutUnit> gridItemOffset, const LayoutUnit totalGuttersSize)
```
Comment (mattwoodrow) [inline Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp]: `computeGridSpanSize`?  A grid area is the 2D rows/columns pairing.

File: Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp [hunk 2]
```diff
@@ -1001,6 +999,10 @@ void GridTrackSizingAlgorithm::cacheBaselineAlignedItem(const RenderBox& item, G
     if (GridLayoutFunctions::isOrthogonalParent(*m_renderGrid, *item.parent()))
         axis = axis == GridColumnAxis ? GridRowAxis : GridColumnAxis;
 
+    const auto* gridItemParent = dynamicDowncast<RenderGrid>(item.parent());
+    if (gridItemParent && axis == GridColumnAxis && !m_renderGrid->isSubgridRows() && gridItemParent->isSubgridOf(GridLayoutFunctions::flowAwareDirectionForChild(*m_renderGrid, item, GridTrackSizingDirection::ForRows), *m_renderGrid))
```
Comment (mattwoodrow) [inline Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp]: I'm finding this condition quite hard to parse and understand.

Is this something where we could pass down a flag through the recursion to say 'we're inside a subgrid of the root, in the axis that the root calls Rows'?

Failing that, maybe just a comment trying to explain what we're catching here.

File: Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp [hunk 3]
```diff
@@ -1570,6 +1572,19 @@ void GridTrackSizingAlgorithm::setup(GridTrackSizingDirection direction, unsigne
     m_hasPercentSizedRowsIndefiniteHeight = false;
     m_hasFlexibleMaxTrackBreadth = false;
 
+    auto resolveAndSetNonAutoRowStartMarginsOnRowSubgrids = [&] {
+        if (m_direction == ForRows) {
```
Comment (mattwoodrow) [inline Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp]: This condition is enforced by the caller.

File: Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp (hunk 1 above)
Comment (sammygill) [reply Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp]: Changed

File: Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp (hunk 2 above)
Comment (sammygill) [reply Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp]: I went ahead and started to keep track of a bool to represent when we want to cache row subgrids and is initially set to true when called on the root grid. We can check to see if that bool is true and if the item's parent is a row subgrid. 

It makes the logic slightly shorter and hopefully easier to read and understand. Let me know what you think!

File: Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp (hunk 3 above)
Comment (sammygill) [reply Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp]: Fixed

**Review by mattwoodrow — APPROVED** [review]
